demos/coreMQTT_Agent/mqtt_agent_task.c

```diff
@@ -0,0 +1,1025 @@
+
+    return ( xNetworkStatus == TRANSPORT_SOCKET_STATUS_SUCCESS ) ? pdPASS : pdFAIL;
+}
+
+/*-----------------------------------------------------------*/
+
+static void prvIncomingPublishCallback( MQTTAgentContext_t * pMqttAgentContext,
+                                        uint16_t packetId,
+                                        MQTTPublishInfo_t * pxPublishInfo )
+{
+    bool xPublishHandled = false;
+    char cOriginalChar, * pcLocation;
+
+    ( void ) packetId;
+
+    /* Fan out the incoming publishes to the callbacks registered using
+     * subscription manager. */
+    xPublishHandled = handleIncomingPublishes( ( SubscriptionElement_t * ) pMqttAgentContext->pIncomingCallbackContext,
+                                               pxPublishInfo );
+
+    /* If there are no callbacks to handle the incoming publishes,
+     * handle it as an unsolicited publish. */
+    if( xPublishHandled != true )
+    {
+        /* Ensure the topic string is terminated for printing.  This will over-
+         * write the message ID, which is restored afterwards. */
+        pcLocation = ( char * ) &( pxPublishInfo->pTopicName[ pxPublishInfo->topicNameLength ] );
+        cOriginalChar = *pcLocation;
+        *pcLocation = 0x00;
+        LogWarn( ( "WARN:  Received an unsolicited publish from topic %s", pxPublishInfo->pTopicName ) );
+        *pcLocation = cOriginalChar;
+    }
+}
+
+/*-----------------------------------------------------------*/
+
+static void prvMQTTAgentTask( void * pvParameters )
+{
+    BaseType_t xNetworkResult = pdFAIL;
+    MQTTStatus_t xMQTTStatus = MQTTSuccess, xConnectStatus = MQTTSuccess;
+    MQTTContext_t * pMqttContext = &( xGlobalMqttAgentContext.mqttContext );
+
+    ( void ) pvParameters;
+
+    do
+    {
+        /* MQTTAgent_CommandLoop() is effectively the agent implementation.  It
+         * will manage the MQTT protocol until such time that an error occurs,
+         * which could be a disconnect.  If an error occurs the MQTT context on
+         * which the error happened is returned so there can be an attempt to
+         * clean up and reconnect however the application writer prefers. */
+        xMQTTStatus = MQTTAgent_CommandLoop( &xGlobalMqttAgentContext );
+
+        /* Success is returned for disconnect or termination. The socket should
+         * be disconnected. */
+        if( ( xMQTTStatus == MQTTSuccess ) && ( xGlobalMqttAgentContext.mqttContext.connectStatus == MQTTNotConnected ) )
+        {
+            /* MQTT Disconnect. Disconnect the socket. */
+            xNetworkResult = prvSocketDisconnect( &xNetworkContext );
+        }
+        else if( xMQTTStatus == MQTTSuccess )
+        {
+            /* MQTTAgent_Terminate() was called, but MQTT was not disconnected. */
+            xConnectStatus = MQTT_Disconnect( &( xGlobalMqttAgentContext.mqttContext ) );
+            xNetworkResult = prvSocketDisconnect( &xNetworkContext );
+            break;
+        }
+        /* Error. */
```

Comment:
What kind of errors does this represent for which connection is retried and the agent resumed?

Reply:
Any error. Any status code that's not `MQTTSuccess`.